### DISCUSSION TITLE
Tensor Mechanics and Heat Conduction coupling

### DISCUSSION THREAD

**Discussion by krb-nantes** [idaholab/moose#22348] [Q&A Modules: Solid mechanics] [answered]
Hi everyone! 
I am currently trying to make a code using Tensor Mechanics Action to calculate the SED for a piece of Zircaloy-4 with a temperature transient. I will add at the end the input for Initial Conditions, Boundary Conditions, TensorMechanics/Master, and Materials block I used for this code.

When I do the simulation using the same mesh, but with only Heat Conduction or only Tensor Mechanics, both work, just when I try to put them together it does not converge anymore.

*Note 1: If I do not use the "use_displaced_mesh = true" it returns a negative temperature in some elements in ZryThermal material.
*Note 2: I tried to add the ZryThermalExpansionMATPROEigenstrain in Material, but it did not change.

If anyone has any idea on how could I solve this problem, I appreciate the help!
```
**INPUT**
[Functions]
####### FUNCTION FOR HEAT CONDUCTION PROBLEM #########
  [temp_IC]
    type = ParsedFunction
    value = 588.70              #-35561*x+643.15 Initial T constant or as a gradient, does not change convergency.
  []
  [inner_temperature_profile]
    type = PiecewiseLinear
    format = columns
    xy_data = '0                622.26
               9616346.107      622.36
               10974285.48      441.15
               11449485.48      293.15'
  []
  [outer_temp_profile]
    type = PiecewiseLinear
    format = columns
    xy_data = '0                588.70
               9616346.107      588.70
               10974285.48      293.15
               11449485.48      293.15'
  []
####### FUNCTION FOR TENSOR MECHANICS #########  
  [inner_disp_profile]
    type = PiecewiseBilinear
    axis = 1
    data_file = inner_disp_x.csv
  []
  [outer_disp_profile]
    type = PiecewiseBilinear
    axis = 1
    data_file = outer_disp_x.csv
  []
[]

[ICs]
    [temp_ic]
        type = FunctionIC
        variable = temp
        function = temp_IC
    []
[]

[BCs]
########### BC FOR HEAT CONDUCTION ##########
  [outer_temp]
    type = FunctionDirichletBC
    variable = temp
    boundary = 'right spallation'
    function = outer_temp_profile
    use_displaced_mesh = true
    preset = false
  []
  [inner_temperature]
    type = FunctionDirichletBC
    variable = temp
    boundary = 'left'
    function = inner_temperature_profile
    use_displaced_mesh = true
    preset = false
  []
########### BC FOR TENSOR MECHANICS ##########
 [inner_disp]
    type = FunctionDirichletBC
    variable = disp_x
    boundary = 'left'
    function = inner_disp_profile
    preset = false
    use_displaced_mesh = true
  []
  [outer_disp]
    type = FunctionDirichletBC
    variable = disp_x
    boundary = 'right'
    function = outer_disp_profile
    preset = false
    use_displaced_mesh = true
  []
  [no_disp_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom'
    value = 0.0
    preset = false
    use_displaced_mesh = true
  []
[]
[Modules/TensorMechanics/Master]
    [tensor]                        # this division is necessary
        strain = SMALL
        incremental = true
        add_variables = true
        generate_output = 'stress_xx stress_yy stress_zz vonmises_stress hoop_stress strain_xx strain_yy strain_zz hoop_strain'
        planar_formulation = PLANE_STRAIN
    []
[]

[Materials]
########### MATERIAL FOR HEAT CONDUCTION #####
  [clad_thermal]
    type = ZryThermal
    block = '0'
    temperature = temp
    use_displaced_mesh = true
  []
  [density_clad]
    type = Density
    block = '0'
    density = 6550                         # kg/m3
  []
  [oxide_conduction]                    # delta-ZrH1.66 @ 500K
    type = GenericConstantMaterial
    prop_values =  '1.675 0'            # W/mK
    prop_names = 'thermal_conductivity thermal_conductivity_dT'
    block = '1'
    use_displaced_mesh = true
  []
  [oxide_density]
    type = GenericConstantMaterial
    prop_values = 6150                  # kg/m3 from AZOMaterials
    block = '1'
    prop_names = 'density'
  []
  [oxide_cp]
    type = GenericConstantMaterial
    prop_values = 540                   # J/kgK
    block = '1'
    prop_names = 'specific_heat'
    use_displaced_mesh = true
  []
########### MATERIAL FOR TENSOR MECHANICS #########
  [strain_energy_density]
    type = StrainEnergyDensity
    incremental = true
    use_displaced_mesh = true
    block = '0 1'
  []
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 99e+9 
    poissons_ratio = 0.3
    block = '0 1'
    use_displaced_mesh = true
  []  
  [elastic_stress]
    type = ComputeMultipleInelasticStress
    inelastic_models = 'isoplas'
    block = '0 1'
  []
  [isoplas]
    type = IsotropicPlasticityStressUpdate
    yield_stress = 381e6 
    hardening_constant = 0.0
    use_displaced_mesh = true
  []
[]
```

```
**CSV FILES**
_outer_disp_x:_
0,0.003736,0.007472,0.011207
0,0,0,0,0
9616346.107,1.29E-06,1.28E-06,1.27E-06,1.27E-06
10974285.48,-1.08E-05,-1.08E-05,-1.08E-05,-1.08E-05
11449485.48,-1.20E-05,-1.20E-05,-1.20E-05,-1.20E-05

_inner_disp_x:_
0,0.003736,0.007472,0.011207
0,0,0,0,0
9616346.107,1.72E-06,1.70E-06,1.69E-06,1.69E-06
10974285.48,-9.20E-06,-9.20E-06,-9.20E-06,-9.20E-06
11449485.48,-1.05E-05,-1.05E-05,-1.05E-05,-1.05E-05
```

**Comment by friedmud** [comment, marked as answer]
What type of simulation are you trying to do?  If you are trying to do a Steady calculation - you may want to use `Transient` instead... and ramp up the temperature / boundary conditions.

You can also try to "decouple" these - but still solve the coupled problem.  Set temperature as an AuxiliaryVariable and then give it a constant value that is reasonable... and couple that in as temperature in your TensorMechanics problem and make sure that solves.  You may just have an unreasonable material property somewhere that depends on temperature.

**Reply by friedmud** [reply]
Also: you have several HUGE constants in your TensorMechanics.  That will make it tough to converge the heat conduction problem (due to the residual / jacobian entries being so much smaller).  You may want to try scaling / automatic scaling.  The scaling parameters are in the `Executioner` block... you can see them at the bottom of the page here: https://mooseframework.inl.gov/source/executioners/Transient.html

**Reply by GiudGiud** [reply]
Also why did you set preset = false in all the BCs? which example input did you work from?

**Reply by friedmud** [reply]
For solid-mechanics it can sometimes be necessary to turn off preset - imagine a DirichletBC for displacement that is large enough to invert elements... that's exactly what would instantly happen if preset is on.

On the other hand - if you're not going to invert elements with your displacement, it is certainly better to use preset.  Double-edged sword.

**Reply by hugary1995** [reply]
Probably off-topic here, @friedmud I think I was the person who said preset=true is good for all my (mechanics) problems several years ago when we consolidated preset and dirichlet BCs. But over the years I've found that preset=false performs better for plasticity problems with a high hardening rate (even when the preset BC isn't inverting elements). My hypothesis is that the preset BC is forcing certain elements into the return mapping when they shouldn't in the equilibriated solution. 

Anyways, preset=true is still a good default by in large. But the scenario I mentioned above may be a collective knowledge we should document at some point...

**Reply by friedmud** [reply]
That's good to know @hugary1995 .  Turning off preset will allow the solver to "solve to" the final solution... moving with everything else in the system.  This can be "softer"... akin to taking small timesteps and letting everything equilibrate during the solve... so I think it makes sense.

**Reply by krb-nantes** [reply]
Hi all, thanks for helping!
1- I worked based on the examples in: _moose/modules/tensor_mechanics/test/tests/strain_energy_density/_ , but I am not actually using it in all the BC only for the axial constraint, it was my mistake posting here with this in all of them. And, indeed depending on the displacement value I set its necessary to use the preset = false.
2- I tried set the constant temperature as an AuxVariable, and it actually did not change;
3- I think the most possible solution here will be the scaling. As I understood from your suggestion, I apply the scaling parameters on Executioner block (and I am using Transient), however, with these parameters I could only scale the displacement variables and it seems to be going, but very slow. So, I was wondering if it is possible to scale the Material properties as well (Young's modulus and Yield Stress), or this is already done in "automatic_scaling", this still not so clear to me?

**Reply by friedmud** [reply]
The automatic scaling scales the "result" of using those large material properties (the residuals and jacobians that are resultant).  When you say "going slow"... what does that mean?  Can you post some of your convergence history?

**Reply by GiudGiud** [reply]
Any update on this @krb-nantes ?

**Reply by krb-nantes** [reply]
Hi all, so in my first try using the scaling, the code stopped because it reached the wall time of 5h, it did not finished, but also did not have any errors. I do not think I have the performance metrics file for this time, but I have this one below. This one correspond for the same file, I ran it a second time and then I had a non-convergency error.

[mesoscale_TM_HC_performance_metrics.csv](https://github.com/idaholab/moose/files/9862682/mesoscale_TM_HC_performance_metrics.csv)

**Reply by GiudGiud** [reply]
Any update on this?

I've also seen some segregate the two solves using a multiapp and do a fixed point iteration. This should be more stable, especially with relaxation, if that s what s needed

**Reply by krb-nantes** [reply]
Hi, so it's actually still not working. My guess is that it may have some issue with my tensor mechanics problem, so I am working on finding the best boundary conditions set up for it, trying different configurations (maybe changing the displacements values, the side sets, or using the strain instead of displacement).

**Reply by GiudGiud** [reply]
sounds good. If you isolate another issue please feel free to open another post. 

A little advice on working through this:
- work on the smallest system as possible, maybe just a rectangular mesh
- use LU preconditioning on this small system
- separate and validate your input for each physics
- output all material properties to exodus
- leverage the automated tools we have: the jacobian analyzer for non linear convergence issues, automatic scaling, etc and our docs such as this page https://mooseframework.inl.gov/moose/application_usage/failed_solves.html
- for displaced meshes, not all quantities should be using the displaced mesh, so if you are using GlobalParams, be mindful of it
